frontend/dockerfile/dockerfile_test.go

```diff
@@ -9425,6 +9426,147 @@ COPY Dockerfile /foo
 	}
 }
 
+func testPlatformWithOSVersion(t *testing.T, sb integration.Sandbox) {
+	// This test cannot be run on Windows currently due to `FROM scratch` and
+	// layer formatting not being supported on Windows.
+	integration.SkipOnPlatform(t, "windows")
+
+	ctx := sb.Context()
+
+	c, err := client.New(ctx, sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	f := getFrontend(t, sb)
+
+	// NOTE: currently "OS" *must* be set to "windows" for this to work.
+	// The platform matchers only do OSVersion comparisons when the OS is set to "windows".
+	p1 := ocispecs.Platform{
+		OS:           "windows",
+		OSVersion:    "1.2.3",
+		Architecture: "bar",
+	}
+	p2 := ocispecs.Platform{
+		OS:           "windows",
+		OSVersion:    "1.1.0",
+		Architecture: "bar",
+	}
+
+	p1Str := platforms.FormatAll(p1)
+	p2Str := platforms.FormatAll(p2)
+
+	registry, err := sb.NewRegistry()
+	if errors.Is(err, integration.ErrRequirements) {
+		t.Skip(err.Error())
+	}
+	require.NoError(t, err)
+	target := registry + "/buildkit/testplatformwithosversion:latest"
+
+	dockerfile := []byte(`
+ARG TARGETOS TARGETOSVERSION TARGETARCH
+FROM --platform=${TARGETOS}(${TARGETOSVERSION})/${TARGETARCH} ` + target + ` AS reg
```

Comment:
This platform flag in here shouldn't be needed for functionality as the default should work the same. `FROM --platform=$TARGETPLATFORM` should also work the same.

If you want to check that the parser works in these cases then maybe add other stages for different variants (reg1, reg2, reg3) and build all in a loop.

Reply:
Your right no need for setting anything here.

Since you brought it up, I did not change `TARGETPLATFORM` or `BUILDPLATFORM` to use `FormatAll` for fear of breaking things.
I guess `TARGETPLATFORM` shouldn't be a big deal since it'd just be what the user specified.
`BUILDPLATFORM` would be different... at least on Windows... which maybe isn't a big deal since its experimental?

Reply:
Yeah, I think `TARGETPLATFORM` is fine. `BUILDPLATFORM` maybe indeed shouldn't set `OSVersion`. This would be different if there would be "default" osversion and different `DefaultSpec()` and `MaximumSpec()`. Eg. in `linux/amd64`, `DefaultSpec()` variant is always `v1`, doesn't matter if the machine actually supports higher versions. In `OSVersion` this concept doesn't seem to exist and it goes to syscall https://github.com/containerd/platforms/blob/main/defaults_windows.go#L29C33-L29C55 . It's possible we need to define our own `DefaultSpec` in the follow-up.